{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d7dec497-e30d-4ba0-aba9-2b4fa51f0732",
   "metadata": {},
   "source": [
    "# Loading all abstracts as text for further processing for models"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0151ec5e-ff8e-4666-b0b9-3f0249ec16b5",
   "metadata": {},
   "source": [
    " **This noebook can be duplicated to use as the beginning to input the relevant dataset on Cellbiology**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aad7fc2c-5041-4f8a-8f15-22a5bb0145ac",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: transformers in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (4.39.3)\n",
      "Requirement already satisfied: filelock in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (3.13.3)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.19.3 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (0.22.2)\n",
      "Requirement already satisfied: numpy>=1.17 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (1.24.3)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (23.2)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (6.0.1)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (2023.12.25)\n",
      "Requirement already satisfied: requests in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (2.31.0)\n",
      "Requirement already satisfied: tokenizers<0.19,>=0.14 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (0.15.2)\n",
      "Requirement already satisfied: safetensors>=0.4.1 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (0.4.2)\n",
      "Requirement already satisfied: tqdm>=4.27 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from transformers) (4.66.2)\n",
      "Requirement already satisfied: fsspec>=2023.5.0 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from huggingface-hub<1.0,>=0.19.3->transformers) (2024.3.1)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from huggingface-hub<1.0,>=0.19.3->transformers) (4.9.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from requests->transformers) (3.3.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from requests->transformers) (3.6)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from requests->transformers) (2.2.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/jennyluise/.pyenv/versions/3.11.3/lib/python3.11/site-packages (from requests->transformers) (2024.2.2)\n"
     ]
    }
   ],
   "source": [
    "!pip install transformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c553fa1f-44ef-44f5-bd96-b38133cf125f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Import required libraries\n",
    "\n",
    "import pandas as pd\n",
    "import csv\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap\n",
    "from tqdm import tqdm\n",
    "import re\n",
    "import os\n",
    "\n",
    "from transformers import pipeline\n",
    "import sklearn\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.feature_selection import SelectKBest, f_classif\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "from sklearn.metrics import accuracy_score, classification_report\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import confusion_matrix\n",
    "import time\n",
    "from time_calc_function import seconds_to_hms\n",
    "\n",
    "import time\n",
    "from time_calc_function import seconds_to_hms \n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "126aead9-eac8-4588-afa7-656ae3ee8d3b",
   "metadata": {},
   "source": [
    "# Load & clean non retracted papers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccf9056e-c17d-4b77-9b3a-b395f9bd79fa",
   "metadata": {},
   "source": [
    "**Reading the csv file with the text of non-retracted papers into a dataframe**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "59de722b-dd9d-4961-8f61-d04ba2fa7d78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Relative path to the CSV file for non-retracted papers\n",
    "file_path = '../data/cellbio_abstracts/cellbio_abstracts_non-retracted_text.csv'\n",
    "\n",
    "# Load the CSV file into a DataFrame, specifying the pipe '|' as the delimiter\n",
    "abstract_nonretracted_df = pd.read_csv(file_path, delimiter='|')\n",
    "\n",
    "# Display information about the DataFrame \n",
    "#abstract_nonretracted_df.info()\n",
    "#abstract_nonretracted_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "810715e5-8b98-4e1f-90c4-5490fc6e0cc2",
   "metadata": {
    "tags": []
   },
   "source": [
    "**Removing all rows where retract% or withdraw% appear in the abstract text**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "81481a14-4669-459d-8a0f-35d27270c7c7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# filter the dataframe to clean up rows that have the word retracted in non-retracted papers\n",
    "\n",
    "# Filter the DataFrame to only include rows where 'ret_flag' is True\n",
    "\n",
    "ret_flag_nonretracted_df = abstract_nonretracted_df[abstract_nonretracted_df['ret_flag'] == True]\n",
    "\n",
    "# Info on the filtered DataFrame\n",
    "#ret_flag_nonretracted_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "609d6e66-c2ac-41b8-a303-29e6f33663f0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Drop rows where 'ret_flag' is True directly from the main DataFrame\n",
    "cleaned_nonretracted_df = abstract_nonretracted_df[abstract_nonretracted_df['ret_flag'] != True]\n",
    "\n",
    "# Display the cleaned DataFrame to verify the rows are dropped\n",
    "#cleaned_nonretracted_df.info()\n",
    "#cleaned_nonretracted_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c08aef85-8532-4e49-85df-07c7ce122b42",
   "metadata": {},
   "source": [
    "# Load & clean retracted papers "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19c3be49-f6cc-4360-b3a6-907ffefd4b00",
   "metadata": {},
   "source": [
    "**Reading the csv file with the text of retracted papaers into a dataframe**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "574efc8f-6c44-4b47-bb58-4b6af76534a2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Relative path to the CSV file\n",
    "file_path = '../data/cellbio_abstracts/cellbio_abstracts_retracted_text.csv'\n",
    "\n",
    "# Load the CSV file into a DataFrame, specifying the pipe '|' as the delimiter\n",
    "abstract_retracted_df = pd.read_csv(file_path, delimiter='|')\n",
    "\n",
    "# Display information about the DataFrame \n",
    "#abstract_retracted_df.info()\n",
    "#abstract_retracted_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bce75bd-1906-4848-b3b7-0d728834639b",
   "metadata": {},
   "source": [
    "**Removing all rows where retract% or withdraw% appear in the abstract text**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "83be343c-767d-46b3-b28e-24aff1843b17",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# filter the dataframe to clean up rows that have the word retracted :\n",
    "\n",
    "# Filter the DataFrame to only include rows where 'ret_flag' is True\n",
    "ret_flag_retracted_df = abstract_retracted_df[abstract_retracted_df['ret_flag'] == True]\n",
    "\n",
    "# Display the filtered DataFrame\n",
    "#ret_flag_retracted_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aa735280-cf99-480f-9b44-5ab71a79977c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Drop rows where 'ret_flag' is True directly from the main DataFrame\n",
    "\n",
    "cleaned_retracted_df = abstract_retracted_df[abstract_retracted_df['ret_flag'] != True]\n",
    "\n",
    "# Display the cleaned DataFrame to verify the rows are dropped\n",
    "#cleaned_retracted_df.info()\n",
    "#cleaned_retracted_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23cd8e38-3c40-4fb3-95fb-b22912ab9f85",
   "metadata": {},
   "source": [
    "# Dataframe with just abstract_text and target for both retracted and non-retracted papers\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c9441ba-a7bc-44d8-b860-ce93f282d80e",
   "metadata": {},
   "source": [
    "Create one Dataframe each target group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7a9a9a61-f7c9-496e-965c-f361416d525f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# Selecting required columns for retracted abstracts\n",
    "retracted_subset = cleaned_retracted_df[['abstract_text', 'target']]\n",
    "\n",
    "# Selecting required columns for non-retracted abstracts\n",
    "nonretracted_subset = cleaned_nonretracted_df[['abstract_text', 'target']]\n",
    "\n",
    "# Creating new dataframes\n",
    "retracted_df = pd.DataFrame(retracted_subset)\n",
    "nonretracted_df = pd.DataFrame(nonretracted_subset)\n",
    "\n",
    "#Display dfs\n",
    "#retracted_df\n",
    "#nonretracted_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "918c91a8-9d76-47c1-a189-fe83cea2ccc2",
   "metadata": {},
   "source": [
    "# Combine both target groups into one dataframe for dataset model processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "41b419de-caf2-408b-b126-cb38b4c1f352",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract_text</th>\n",
       "      <th>target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>\"Abstract Glioblastomas are lethal brain tumor...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>\" h2 Summary /h2  Septins are GTP-binding prot...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>\"Genome-wide association studies (GWAS) have i...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>\"Coupling of post-Golgi and endocytic membrane...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>\"Abstract Objective: To explore the correlatio...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12698</th>\n",
       "      <td>\"Abstract Selforganization is a fundamental an...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12699</th>\n",
       "      <td>\"Diaporthe species have often been reported as...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12700</th>\n",
       "      <td>\"The zebrafish has emerged as a powerful new v...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12701</th>\n",
       "      <td>\"Drebrin is an actin bundling protein that pla...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12702</th>\n",
       "      <td>\"A growing body of multiscale biomechanical st...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12703 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           abstract_text  target\n",
       "0      \"Abstract Glioblastomas are lethal brain tumor...       1\n",
       "1      \" h2 Summary /h2  Septins are GTP-binding prot...       0\n",
       "2      \"Genome-wide association studies (GWAS) have i...       0\n",
       "3      \"Coupling of post-Golgi and endocytic membrane...       0\n",
       "4      \"Abstract Objective: To explore the correlatio...       1\n",
       "...                                                  ...     ...\n",
       "12698  \"Abstract Selforganization is a fundamental an...       0\n",
       "12699  \"Diaporthe species have often been reported as...       0\n",
       "12700  \"The zebrafish has emerged as a powerful new v...       0\n",
       "12701  \"Drebrin is an actin bundling protein that pla...       0\n",
       "12702  \"A growing body of multiscale biomechanical st...       0\n",
       "\n",
       "[12703 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Concatenating both dataframes\n",
    "abstracts_df = pd.concat([retracted_df, nonretracted_df], ignore_index=True)\n",
    "\n",
    "# Optionally, you can shuffle the rows if you want\n",
    "abstracts_df = abstracts_df.sample(frac=1).reset_index(drop=True)\n",
    "\n",
    "#make sure target is a integer\n",
    "abstracts_df['target'] = abstracts_df['target'].astype(int)\n",
    "\n",
    "\n",
    "# Print the combined dataframe\n",
    "abstracts_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0b0192e1-3d47-41d2-9ee8-db26806a9ea9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 12703 rows in the dataset\n"
     ]
    }
   ],
   "source": [
    "print(f'There are {abstracts_df.shape[0]} rows in the dataset')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4e9c6a33-6b7a-4e81-9a37-5e2b111ea53f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# split Y and y\n",
    "\n",
    "X = abstracts_df[\"abstract_text\"]\n",
    "y = abstracts_df[\"target\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "25fa9023-c13d-47fa-bd2d-bd5996f31deb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        \"Abstract Glioblastomas are lethal brain tumor...\n",
       "1        \" h2 Summary /h2  Septins are GTP-binding prot...\n",
       "2        \"Genome-wide association studies (GWAS) have i...\n",
       "3        \"Coupling of post-Golgi and endocytic membrane...\n",
       "4        \"Abstract Objective: To explore the correlatio...\n",
       "                               ...                        \n",
       "12698    \"Abstract Selforganization is a fundamental an...\n",
       "12699    \"Diaporthe species have often been reported as...\n",
       "12700    \"The zebrafish has emerged as a powerful new v...\n",
       "12701    \"Drebrin is an actin bundling protein that pla...\n",
       "12702    \"A growing body of multiscale biomechanical st...\n",
       "Name: abstract_text, Length: 12703, dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9e20a1d9-92ae-40a1-a2a4-82d73fefbf53",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# Split data into train and text sets\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "262929b0-a736-499b-b221-bc5914b6501f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2629     \"Vitiligo is associated with oxidant stress an...\n",
       "10456    \"In this work  silver nanoparticles were synth...\n",
       "7136     \"Case series have demonstrated that potentiall...\n",
       "8546     \"Mesenchymal stem cells (MSCs) can sense and c...\n",
       "12401    \"As a core component of the mitotic checkpoint...\n",
       "                               ...                        \n",
       "11964    \"Colorectal cancer (CRC) is the third most com...\n",
       "5191     \"MicroRNA-29 (miR-29) family is involved in va...\n",
       "5390     \"Lophiostoma bipolare was taxonomically revise...\n",
       "860      \"The aim of our study was to detect the expres...\n",
       "7270     \"Spatial and functional organization of cells ...\n",
       "Name: abstract_text, Length: 10162, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeb99215-7fdb-4fc1-89b6-1121b9448a3a",
   "metadata": {},
   "source": [
    "**Create BOW Matrix for trianing dataset abstracts using Tf-idf vecorizer removing stoppwords only**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e8da4d80-0242-434d-b558-4f8c7785ca1e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\n# Initialize TfidfVectorizer with stopwords removed\\n\\ntfidf_vectorizer = TfidfVectorizer(stop_words='english')\\n\\n# Fit and transform the abstract_text column\\n\\ntrain_tfidf_matrix = tfidf_vectorizer.fit_transform(X_train)\\n\\n# Convert the TF-IDF matrix to an array\\n\\ntrain_tfidf_matrix_array = train_tfidf_matrix.toarray()\\n\\n# Convert the array to a DataFrame if needed\\n\\ntrain_tfidf_df = pd.DataFrame(train_tfidf_matrix_array, columns=tfidf_vectorizer.get_feature_names_out())\\n\\n# Output the DataFrame\\ntrain_tfidf_df\""
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "# Initialize TfidfVectorizer with stopwords removed\n",
    "\n",
    "tfidf_vectorizer = TfidfVectorizer(stop_words='english')\n",
    "\n",
    "# Fit and transform the abstract_text column\n",
    "\n",
    "train_tfidf_matrix = tfidf_vectorizer.fit_transform(X_train)\n",
    "\n",
    "# Convert the TF-IDF matrix to an array\n",
    "\n",
    "train_tfidf_matrix_array = train_tfidf_matrix.toarray()\n",
    "\n",
    "# Convert the array to a DataFrame if needed\n",
    "\n",
    "train_tfidf_df = pd.DataFrame(train_tfidf_matrix_array, columns=tfidf_vectorizer.get_feature_names_out())\n",
    "\n",
    "# Output the DataFrame\n",
    "train_tfidf_df\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "139c1d38-57ca-4252-9f65-2f12dfe015f2",
   "metadata": {},
   "source": [
    "**Create BOW Matrix for test dataset abstracts using Tf-idf vecorizer removing stoppwords only**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "08623e2a-5ad9-4555-9546-0d160a6ea4ca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# vectorize X for test dataset with  TfidfVectorizer with stopwords removed as initialized above\\n\\n\\n# Fit and transform the abstract_text column\\ntest_tfidf_matrix = tfidf_vectorizer.fit_transform(X_test)\\n\\n# Convert the TF-IDF matrix to an array\\ntest_tfidf_matrix_array = test_tfidf_matrix.toarray()\\n\\n# Convert the array to a DataFrame if needed\\ntest_tfidf_df = pd.DataFrame(test_tfidf_matrix_array, columns=tfidf_vectorizer.get_feature_names_out())\\n\\n# Output the DataFrame\\ntest_tfidf_df'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"# vectorize X for test dataset with  TfidfVectorizer with stopwords removed as initialized above\n",
    "\n",
    "\n",
    "# Fit and transform the abstract_text column\n",
    "test_tfidf_matrix = tfidf_vectorizer.fit_transform(X_test)\n",
    "\n",
    "# Convert the TF-IDF matrix to an array\n",
    "test_tfidf_matrix_array = test_tfidf_matrix.toarray()\n",
    "\n",
    "# Convert the array to a DataFrame if needed\n",
    "test_tfidf_df = pd.DataFrame(test_tfidf_matrix_array, columns=tfidf_vectorizer.get_feature_names_out())\n",
    "\n",
    "# Output the DataFrame\n",
    "test_tfidf_df\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "403dbda9-6b6c-470b-869f-8e32f6241b4d",
   "metadata": {},
   "source": [
    "**Alternative approach: Create BOW Matrix for training and test dataset abstracts using Tf-idf vecorizer removing stoppwords and numbers**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "39789b60-4526-42c1-bb98-d8fd468aaf92",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA</th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAACAGAGTAGCAGCGCAGACTGC</th>\n",
       "      <th>AAAGCCTATCATGGAGAAGAGGCG</th>\n",
       "      <th>AAAUGCCCUGUCUUACUGUCA</th>\n",
       "      <th>AAAs</th>\n",
       "      <th>AACAT</th>\n",
       "      <th>AACCT</th>\n",
       "      <th>AACR</th>\n",
       "      <th>AACTAAA</th>\n",
       "      <th>...</th>\n",
       "      <th>zygote</th>\n",
       "      <th>zygotes</th>\n",
       "      <th>zygotic</th>\n",
       "      <th>zyjingfz</th>\n",
       "      <th>zymogen</th>\n",
       "      <th>zymography</th>\n",
       "      <th>zymolyase</th>\n",
       "      <th>zyxin</th>\n",
       "      <th>zyy</th>\n",
       "      <th>zzu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10157</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10158</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10159</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10160</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10161</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10162 rows × 62565 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        AA  AAA  AAACAGAGTAGCAGCGCAGACTGC  AAAGCCTATCATGGAGAAGAGGCG   \n",
       "0      0.0  0.0                       0.0                       0.0  \\\n",
       "1      0.0  0.0                       0.0                       0.0   \n",
       "2      0.0  0.0                       0.0                       0.0   \n",
       "3      0.0  0.0                       0.0                       0.0   \n",
       "4      0.0  0.0                       0.0                       0.0   \n",
       "...    ...  ...                       ...                       ...   \n",
       "10157  0.0  0.0                       0.0                       0.0   \n",
       "10158  0.0  0.0                       0.0                       0.0   \n",
       "10159  0.0  0.0                       0.0                       0.0   \n",
       "10160  0.0  0.0                       0.0                       0.0   \n",
       "10161  0.0  0.0                       0.0                       0.0   \n",
       "\n",
       "       AAAUGCCCUGUCUUACUGUCA  AAAs  AACAT  AACCT  AACR  AACTAAA  ...  zygote   \n",
       "0                        0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0  \\\n",
       "1                        0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "2                        0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "3                        0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "4                        0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "...                      ...   ...    ...    ...   ...      ...  ...     ...   \n",
       "10157                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "10158                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "10159                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "10160                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "10161                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "\n",
       "       zygotes  zygotic  zyjingfz  zymogen  zymography  zymolyase  zyxin  zyy   \n",
       "0          0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0  \\\n",
       "1          0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "2          0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "3          0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "4          0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "...        ...      ...       ...      ...         ...        ...    ...  ...   \n",
       "10157      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "10158      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "10159      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "10160      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "10161      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "\n",
       "       zzu  \n",
       "0      0.0  \n",
       "1      0.0  \n",
       "2      0.0  \n",
       "3      0.0  \n",
       "4      0.0  \n",
       "...    ...  \n",
       "10157  0.0  \n",
       "10158  0.0  \n",
       "10159  0.0  \n",
       "10160  0.0  \n",
       "10161  0.0  \n",
       "\n",
       "[10162 rows x 62565 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Removing numbers reduces the columns from 68k to 63k for the cellbio dataset. Used as alternative\n",
    "# Function to remove numbers\n",
    "def remove_numbers(text):\n",
    "    return re.sub(r'\\d+', '', text)\n",
    "\n",
    "\n",
    "# Initialize TfidfVectorizer with stopwords and numbers removed\n",
    "tfidf_vectorizer = TfidfVectorizer(stop_words='english', preprocessor=remove_numbers)\n",
    "\n",
    "# Fit and transform the abstract_text column\n",
    "train_tfidf_matrix = tfidf_vectorizer.fit_transform(X_train)\n",
    "\n",
    "# Convert the TF-IDF matrix to an array\n",
    "train_tfidf_matrix_array = train_tfidf_matrix.toarray()\n",
    "\n",
    "# Convert the array to a DataFrame if needed\n",
    "train_tfidf_df = pd.DataFrame(train_tfidf_matrix_array, columns=tfidf_vectorizer.get_feature_names_out())\n",
    "\n",
    "# Output the DataFrame\n",
    "train_tfidf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f1ee7ae7-1bcc-4e44-8e2c-e8d2150a386b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AA</th>\n",
       "      <th>AAA</th>\n",
       "      <th>AAACAGAGTAGCAGCGCAGACTGC</th>\n",
       "      <th>AAAGCCTATCATGGAGAAGAGGCG</th>\n",
       "      <th>AAAUGCCCUGUCUUACUGUCA</th>\n",
       "      <th>AAAs</th>\n",
       "      <th>AACAT</th>\n",
       "      <th>AACCT</th>\n",
       "      <th>AACR</th>\n",
       "      <th>AACTAAA</th>\n",
       "      <th>...</th>\n",
       "      <th>zygote</th>\n",
       "      <th>zygotes</th>\n",
       "      <th>zygotic</th>\n",
       "      <th>zyjingfz</th>\n",
       "      <th>zymogen</th>\n",
       "      <th>zymography</th>\n",
       "      <th>zymolyase</th>\n",
       "      <th>zyxin</th>\n",
       "      <th>zyy</th>\n",
       "      <th>zzu</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2536</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2537</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2538</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2539</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2540</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2541 rows × 62565 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       AA  AAA  AAACAGAGTAGCAGCGCAGACTGC  AAAGCCTATCATGGAGAAGAGGCG   \n",
       "0     0.0  0.0                       0.0                       0.0  \\\n",
       "1     0.0  0.0                       0.0                       0.0   \n",
       "2     0.0  0.0                       0.0                       0.0   \n",
       "3     0.0  0.0                       0.0                       0.0   \n",
       "4     0.0  0.0                       0.0                       0.0   \n",
       "...   ...  ...                       ...                       ...   \n",
       "2536  0.0  0.0                       0.0                       0.0   \n",
       "2537  0.0  0.0                       0.0                       0.0   \n",
       "2538  0.0  0.0                       0.0                       0.0   \n",
       "2539  0.0  0.0                       0.0                       0.0   \n",
       "2540  0.0  0.0                       0.0                       0.0   \n",
       "\n",
       "      AAAUGCCCUGUCUUACUGUCA  AAAs  AACAT  AACCT  AACR  AACTAAA  ...  zygote   \n",
       "0                       0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0  \\\n",
       "1                       0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "2                       0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "3                       0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "4                       0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "...                     ...   ...    ...    ...   ...      ...  ...     ...   \n",
       "2536                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "2537                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "2538                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "2539                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "2540                    0.0   0.0    0.0    0.0   0.0      0.0  ...     0.0   \n",
       "\n",
       "      zygotes  zygotic  zyjingfz  zymogen  zymography  zymolyase  zyxin  zyy   \n",
       "0         0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0  \\\n",
       "1         0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "2         0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "3         0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "4         0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "...       ...      ...       ...      ...         ...        ...    ...  ...   \n",
       "2536      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "2537      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "2538      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "2539      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "2540      0.0      0.0       0.0      0.0         0.0        0.0    0.0  0.0   \n",
       "\n",
       "      zzu  \n",
       "0     0.0  \n",
       "1     0.0  \n",
       "2     0.0  \n",
       "3     0.0  \n",
       "4     0.0  \n",
       "...   ...  \n",
       "2536  0.0  \n",
       "2537  0.0  \n",
       "2538  0.0  \n",
       "2539  0.0  \n",
       "2540  0.0  \n",
       "\n",
       "[2541 rows x 62565 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# vectorize X for test dataset with  TfidfVectorizer with stopwords removed as initialized above\n",
    "\n",
    "# Fit and transform the abstract_text column\n",
    "test_tfidf_matrix = tfidf_vectorizer.transform(X_test)\n",
    "\n",
    "# Convert the TF-IDF matrix to an array\n",
    "test_tfidf_matrix_array = test_tfidf_matrix.toarray()\n",
    "\n",
    "# Convert the array to a DataFrame if needed\n",
    "test_tfidf_df = pd.DataFrame(test_tfidf_matrix_array, columns=tfidf_vectorizer.get_feature_names_out())\n",
    "\n",
    "# Output the DataFrame\n",
    "test_tfidf_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68ab5c1c-c6fe-4c91-8cf3-407e9e093a24",
   "metadata": {},
   "source": [
    "**Configure and rund Random forrest "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "562bba55-75ef-4336-9dae-74d7e4c43b5f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training completed in: 0.0 hours, 5.0 minutes, and 51.42 seconds\n",
      "Prediction completed in: 0.0 hours, 0.0 minutes, and 1.87 seconds\n",
      "Accuracy: 0.7701692247146793\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Configuration of the RandomForest Classifier\n",
    "clf = RandomForestClassifier(\n",
    "    n_estimators=8000,\n",
    "    max_depth=20,\n",
    "    max_features='sqrt',\n",
    "    random_state=42,\n",
    "    n_jobs=-1\n",
    ")\n",
    "\n",
    "# Start timing for fitting the model\n",
    "start_time = time.time()\n",
    "clf.fit(train_tfidf_df, y_train)\n",
    "end_time = time.time()\n",
    "fit_time = end_time - start_time\n",
    "hours, minutes, seconds = seconds_to_hms(fit_time)\n",
    "print(f\"Training completed in: {hours} hours, {minutes} minutes, and {seconds:.2f} seconds\")\n",
    "\n",
    "# Start timing for making predictions\n",
    "start_time = time.time()\n",
    "y_pred = clf.predict(test_tfidf_df)\n",
    "end_time = time.time()\n",
    "predict_time = end_time - start_time\n",
    "hours, minutes, seconds = seconds_to_hms(predict_time)\n",
    "print(f\"Prediction completed in: {hours} hours, {minutes} minutes, and {seconds:.2f} seconds\")\n",
    "\n",
    "# Evaluate the model\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "print(\"Accuracy:\", accuracy)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2d82080b-3413-40da-8436-b674ce54b386",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miR: 0.0178\n",
      "expression: 0.0124\n",
      "assay: 0.0094\n",
      "reticulum: 0.0080\n",
      "cells: 0.0077\n",
      "invasion: 0.0074\n",
      "apoptosis: 0.0072\n",
      "YAP: 0.0071\n",
      "endoplasmic: 0.0071\n",
      "luciferase: 0.0069\n",
      "proliferation: 0.0068\n",
      "ER: 0.0066\n",
      "Hippo: 0.0063\n",
      "cell: 0.0062\n",
      "blot: 0.0057\n",
      "lines: 0.0055\n",
      "reporter: 0.0054\n",
      "stress: 0.0053\n",
      "microRNA: 0.0053\n",
      "Western: 0.0053\n"
     ]
    }
   ],
   "source": [
    "# Get feature importances from the trained model\n",
    "importances = clf.feature_importances_\n",
    "\n",
    "# Get feature names from the vectorizer\n",
    "feature_names = tfidf_vectorizer.get_feature_names_out()\n",
    "\n",
    "# Define the number of top features you want to display\n",
    "K = 20\n",
    "\n",
    "# Get the indices of the K most important features in descending order of importance\n",
    "indices = importances.argsort()[-K:][::-1]\n",
    "\n",
    "# Display the top K important features with their names and importance scores\n",
    "top_features = [(feature_names[i], importances[i]) for i in indices]\n",
    "for feature, importance in top_features:\n",
    "    print(f\"{feature}: {importance:.4f}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2b55600a-1242-45d0-8c9a-17e3fbe11ae4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# Define function to fit & rpedict random forest\n",
    "\n",
    "\n",
    "def train_and_log_rf(n_estimators, max_depth, train_data, train_labels, test_data, test_labels):\n",
    "    # Configuration of the RandomForest Classifier\n",
    "    clf = RandomForestClassifier(\n",
    "        n_estimators=n_estimators,\n",
    "        max_depth=max_depth,\n",
    "        max_features='sqrt',\n",
    "        random_state=42,\n",
    "        n_jobs=-1\n",
    "    )\n",
    "\n",
    "    # Start timing for fitting the model\n",
    "    start_time = time.time()\n",
    "    clf.fit(train_data, train_labels)\n",
    "    fit_time = time.time() - start_time  # Time in seconds\n",
    "    print(f\"Training Time: {fit_time:.2f} seconds\")\n",
    "\n",
    "    # Start timing for making predictions\n",
    "    start_time = time.time()\n",
    "    y_pred = clf.predict(test_data)\n",
    "    predict_time = time.time() - start_time  # Time in seconds\n",
    "    print(f\"Prediction Time: {predict_time:.2f} seconds\")\n",
    "\n",
    "    # Evaluate the model\n",
    "    accuracy = accuracy_score(test_labels, y_pred)\n",
    "    print(\"Accuracy:\", accuracy)\n",
    "\n",
    "    # Prepare data for logging\n",
    "    log_data = {\n",
    "        'n_estimators': n_estimators,\n",
    "        'max_depth': max_depth,\n",
    "        'training_time': f\"{fit_time:.2f} seconds\",\n",
    "        'prediction_time': f\"{predict_time:.2f} seconds\",\n",
    "        'accuracy': accuracy\n",
    "    }\n",
    "    log_df = pd.DataFrame([log_data])\n",
    "\n",
    "    # Define log file path\n",
    "    log_file_path = '../data/model_logs/model_performance_log.csv'\n",
    "\n",
    "    # Check if the directory exists, if not, create it\n",
    "    os.makedirs(os.path.dirname(log_file_path), exist_ok=True)\n",
    "\n",
    "    # Check if file exists, append if it does, otherwise write a new file\n",
    "    if os.path.isfile(log_file_path):\n",
    "        log_df.to_csv(log_file_path, mode='a', header=False, index=False)\n",
    "    else:\n",
    "        log_df.to_csv(log_file_path, mode='w', header=True, index=False)\n",
    "\n",
    "    # Return the model, predictions, and accuracy\n",
    "    return clf, y_pred, accuracy\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "08ea0139-4a8f-4f67-af55-12061dd41604",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training Time: 296.73 seconds\n",
      "Prediction Time: 1.42 seconds\n",
      "Accuracy: 0.8107044470680834\n"
     ]
    }
   ],
   "source": [
    "# Define number of trees & tree depth and instatiate fitting the model and making the predicitons\n",
    "\n",
    "num_trees = 4000\n",
    "tree_depth = 40\n",
    "\n",
    "clf, y_pred, accuracy = train_and_log_rf(num_trees, tree_depth, train_tfidf_df, y_train, test_tfidf_df, y_test)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d48ad215-53cc-4b5d-aa7e-06c709fc60aa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.81\n",
      "Classification Report:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.79      0.87      0.83      1316\n",
      "           1       0.84      0.75      0.79      1225\n",
      "\n",
      "    accuracy                           0.81      2541\n",
      "   macro avg       0.81      0.81      0.81      2541\n",
      "weighted avg       0.81      0.81      0.81      2541\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Calculate accuracy just for reference\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "print(f\"Accuracy: {accuracy:.2f}\")\n",
    "\n",
    "# Generate and print classification report\n",
    "report = classification_report(y_test, y_pred)\n",
    "print(\"Classification Report:\")\n",
    "print(report)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "31eb4c67-009a-4929-a351-5fbeda6b1cac",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcIAAAE8CAYAAABTgLEsAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAq4klEQVR4nO3deViUVf8/8PewDfsmKKAIiorgAu6Poiy5YC5hWm6PCeSS5paImaUJmNGjIq5pWS4PaU+LqSUWmGhqouICrikgLqm4oKLsOJzfH/6YryOgMzrDpPf7dV1zXd3nPnPuz5mEN/c6MiGEABERkUQZ6LsAIiIifWIQEhGRpDEIiYhI0hiEREQkaQxCIiKSNAYhERFJGoOQiIgkjUFIRESSxiAkIiJJYxASaUlmZiZ69eoFGxsbyGQybNmyRavjX7hwATKZDOvWrdPquC+ywMBABAYG6rsMesExCOmlkp2djXfeeQeNGzeGqakprK2t4efnhyVLlqC4uFin2w4NDcWJEycwb948JCQkoH379jrdXm0KCwuDTCaDtbV1tZ9jZmYmZDIZZDIZFi5cqPH4V69eRVRUFNLT07VQLZFmjPRdAJG2JCYm4s0334RcLsfIkSPRsmVLlJWVYd++fZg+fTpOnTqFL7/8UifbLi4uRmpqKj766CNMnDhRJ9twc3NDcXExjI2NdTL+0xgZGaGoqAi//PILBg8erLJuw4YNMDU1RUlJyTONffXqVURHR8Pd3R2+vr5qvy85OfmZtkf0KAYhvRRycnIwdOhQuLm5ISUlBc7Ozsp1EyZMQFZWFhITE3W2/Zs3bwIAbG1tdbYNmUwGU1NTnY3/NHK5HH5+fvj222+rBOHGjRvRt29fbNq0qVZqKSoqgrm5OUxMTGple/Ry46FReinMnz8fBQUF+Prrr1VCsFKTJk0wZcoU5fKDBw8wd+5ceHh4QC6Xw93dHR9++CFKS0tV3ufu7o5+/fph37596NixI0xNTdG4cWP897//VfaJioqCm5sbAGD69OmQyWRwd3cH8PCQYuV/PyoqKgoymUylbceOHejatStsbW1haWkJT09PfPjhh8r1NZ0jTElJQbdu3WBhYQFbW1uEhITgzJkz1W4vKysLYWFhsLW1hY2NDcLDw1FUVFTzB/uY4cOH49dff8Xdu3eVbWlpacjMzMTw4cOr9L99+zYiIyPRqlUrWFpawtraGq+++ioyMjKUfXbv3o0OHToAAMLDw5WHWCvnGRgYiJYtW+LIkSPw9/eHubm58nN5/BxhaGgoTE1Nq8w/ODgYdnZ2uHr1qtpzJelgENJL4ZdffkHjxo3RpUsXtfqPHj0aH3/8Mdq2bYv4+HgEBAQgNjYWQ4cOrdI3KysLb7zxBnr27Im4uDjY2dkhLCwMp06dAgAMHDgQ8fHxAIBhw4YhISEBixcv1qj+U6dOoV+/figtLUVMTAzi4uLw2muv4c8//3zi+37//XcEBwfjxo0biIqKQkREBPbv3w8/Pz9cuHChSv/Bgwfj/v37iI2NxeDBg7Fu3TpER0erXefAgQMhk8nw008/Kds2btyI5s2bo23btlX6nz9/Hlu2bEG/fv2waNEiTJ8+HSdOnEBAQIAylLy8vBATEwMAGDt2LBISEpCQkAB/f3/lOHl5eXj11Vfh6+uLxYsXIygoqNr6lixZAkdHR4SGhkKhUAAAvvjiCyQnJ2PZsmVwcXFRe64kIYLoBZefny8AiJCQELX6p6enCwBi9OjRKu2RkZECgEhJSVG2ubm5CQBiz549yrYbN24IuVwupk2bpmzLyckRAMSCBQtUxgwNDRVubm5VapgzZ4549McvPj5eABA3b96sse7Kbaxdu1bZ5uvrK+rWrSvy8vKUbRkZGcLAwECMHDmyyvbefvttlTFff/11UadOnRq3+eg8LCwshBBCvPHGG6J79+5CCCEUCoVwcnIS0dHR1X4GJSUlQqFQVJmHXC4XMTExyra0tLQqc6sUEBAgAIhVq1ZVuy4gIEClLSkpSQAQn3zyiTh//rywtLQUAwYMeOocSbq4R0gvvHv37gEArKys1Oq/fft2AEBERIRK+7Rp0wCgyrlEb29vdOvWTbns6OgIT09PnD9//plrflzlucWtW7eioqJCrfdcu3YN6enpCAsLg729vbK9devW6Nmzp3Kejxo3bpzKcrdu3ZCXl6f8DNUxfPhw7N69G7m5uUhJSUFubm61h0WBh+cVDQwe/ppRKBTIy8tTHvY9evSo2tuUy+UIDw9Xq2+vXr3wzjvvICYmBgMHDoSpqSm++OILtbdF0sMgpBeetbU1AOD+/ftq9b948SIMDAzQpEkTlXYnJyfY2tri4sWLKu0NGzasMoadnR3u3LnzjBVXNWTIEPj5+WH06NGoV68ehg4diu+///6JoVhZp6enZ5V1Xl5euHXrFgoLC1XaH5+LnZ0dAGg0lz59+sDKygrfffcdNmzYgA4dOlT5LCtVVFQgPj4eTZs2hVwuh4ODAxwdHXH8+HHk5+ervc369etrdGHMwoULYW9vj/T0dCxduhR169ZV+70kPQxCeuFZW1vDxcUFJ0+e1Oh9j1+sUhNDQ8Nq24UQz7yNyvNXlczMzLBnzx78/vvveOutt3D8+HEMGTIEPXv2rNL3eTzPXCrJ5XIMHDgQ69evx+bNm2vcGwSATz/9FBEREfD398c333yDpKQk7NixAy1atFB7zxd4+Plo4tixY7hx4wYA4MSJExq9l6SHQUgvhX79+iE7OxupqalP7evm5oaKigpkZmaqtF+/fh13795VXgGqDXZ2dipXWFZ6fK8TAAwMDNC9e3csWrQIp0+fxrx585CSkoJdu3ZVO3ZlnWfPnq2y7q+//oKDgwMsLCyebwI1GD58OI4dO4b79+9Xe4FRpR9//BFBQUH4+uuvMXToUPTq1Qs9evSo8pmo+0eJOgoLCxEeHg5vb2+MHTsW8+fPR1pamtbGp5cPg5BeCu+//z4sLCwwevRoXL9+vcr67OxsLFmyBMDDQ3sAqlzZuWjRIgBA3759tVaXh4cH8vPzcfz4cWXbtWvXsHnzZpV+t2/frvLeyhvLH7+lo5KzszN8fX2xfv16lWA5efIkkpOTlfPUhaCgIMydOxfLly+Hk5NTjf0MDQ2r7G3+8MMPuHLlikpbZWBX90eDpmbMmIFLly5h/fr1WLRoEdzd3REaGlrj50jEG+rppeDh4YGNGzdiyJAh8PLyUnmyzP79+/HDDz8gLCwMAODj44PQ0FB8+eWXuHv3LgICAnDo0CGsX78eAwYMqPHS/GcxdOhQzJgxA6+//jomT56MoqIirFy5Es2aNVO5WCQmJgZ79uxB37594ebmhhs3buDzzz9HgwYN0LVr1xrHX7BgAV599VV07twZo0aNQnFxMZYtWwYbGxtERUVpbR6PMzAwwKxZs57ar1+/foiJiUF4eDi6dOmCEydOYMOGDWjcuLFKPw8PD9ja2mLVqlWwsrKChYUFOnXqhEaNGmlUV0pKCj7//HPMmTNHeTvH2rVrERgYiNmzZ2P+/PkajUcSoeerVom06ty5c2LMmDHC3d1dmJiYCCsrK+Hn5yeWLVsmSkpKlP3Ky8tFdHS0aNSokTA2Nhaurq5i5syZKn2EeHj7RN++fats5/HL9mu6fUIIIZKTk0XLli2FiYmJ8PT0FN98802V2yd27twpQkJChIuLizAxMREuLi5i2LBh4ty5c1W28fgtBr///rvw8/MTZmZmwtraWvTv31+cPn1apU/l9h6/PWPt2rUCgMjJyanxMxVC9faJmtR0+8S0adOEs7OzMDMzE35+fiI1NbXa2x62bt0qvL29hZGRkco8AwICRIsWLard5qPj3Lt3T7i5uYm2bduK8vJylX5Tp04VBgYGIjU19YlzIGmSCaHBWXIiIqKXDM8REhGRpDEIiYhI0hiEREQkaQxCIiKSNAYhERFJGoOQiIgkjUFIRESS9lI+WcaszUR9l0Ckd3fSluu7BCK9M1Uj5bhHSEREksYgJCIiSWMQEhGRpDEIiYhI0hiEREQkaQxCIiKSNAYhERFJGoOQiIgkjUFIRESSxiAkIiJJYxASEZGkMQiJiEjSGIRERCRpDEIiIpI0BiEREUkag5CIiCSNQUhERJLGICQiIkljEBIRkaQxCImISNIYhEREJGkMQiIikjQGIRERSRqDkIiIJI1BSEREksYgJCIiSWMQEhGRpDEIiYhI0hiEREQkaQxCIiKSNAYhERFJGoOQiIgkjUFIRESSxiAkIiJJYxASEZGkMQiJiEjSGIRERCRpDEIiIpI0BiEREUkag5CIiCSNQUhERJLGICQiIkljEBIRkaQxCImISNIYhEREJGkMQiIikjQGIRERSRqDkIiIJI1BSEREksYgJCIiSWMQEhGRpDEIqVp+bT3w4+J3cD55HoqPLUf/wNYq60Ne8cEvn0/A37v+g+Jjy9G6Wf0njrdl+fgq47RqVh/rY8OQ+etc3E5dhGObZmHCsEBdTIdIK44cTsOkd8ehR2BX+LTwRMrO31XW+7TwrPa1bs1XVcYqKyvD4IEh8Gnhib/OnKmtKVA1GIRULQszOU6cu4L3Yr+rdr25mQn2p2dj1tItTx1r0r+DIETV9jZerrh5+z7CZ61H2zfm4T9fJyFm0msYN8T/Oasn0o3i4iJ4enpi5qw51a7fuXufyiv6k08hk8nQo2dwlb7xcfPhWLeurksmNRjpuwD6Z0r+8zSS/zxd4/pvE9MAAA2d7Z84Tutm9THlrVfg9+/5uPB7rMq6/249oLJ84UoeOrVuhJBXfLDquz3PWDmR7nTtFoCu3QJqXO/g6KiyvDtlJzp07IQGrq4q7fv2/oHU/X8iLn4Z9u3lv3V902sQ3rp1C2vWrEFqaipyc3MBAE5OTujSpQvCwsLg+Ng/KnqxmJkaY11sGN777Htcz7uv1ntsLE1x516Rjisj0r28W7ewd88fmDvvsyrt0XNmY/HSFTA1M9VTdfQovR0aTUtLQ7NmzbB06VLY2NjA398f/v7+sLGxwdKlS9G8eXMcPnz4qeOUlpbi3r17Ki9RoaiFGdDTzJ82CAcycrBt9wm1+v/LpxHe6NUOX2/6U8eVEenez1s3w9zcAt179lK2CSEw+6MP8ObgoWjRspUeq6NH6W2PcNKkSXjzzTexatUqyGQylXVCCIwbNw6TJk1CamrqE8eJjY1FdHS0SpthvQ4wdu6o9ZpJfX0DWiGwYzP8a+hnT+8MwNvDGd/Hj8W8L7dj54G/dFwdke5t2bwJffr1h1wuV7Zt3JCAwsJCjBrzjh4ro8fpLQgzMjKwbt26KiEIADKZDFOnTkWbNm2eOs7MmTMRERGh0la32wyt1UnPJrBDMzRu4IDcPQtU2r9dOBp/HstG8JglyrbmjZ2w/YtJWLNpP/7zVVJtl0qkdUePHMaFnBzMX7hYpT3t4AEcz0hHhzaqe4PDhwxCn7798Unsf2qxSqqktyB0cnLCoUOH0Lx582rXHzp0CPXq1XvqOHK5XOUvLgCQGRhqpUZ6dgvXJmPt5v0qbUd+/Ajvx21C4h8nlW1ejZ3w65eTseGXg4ha8Uttl0mkE5s3/QjvFi3g+djvtxkzZ2HC5PeUyzdv3MD4saMwf2E8WrX2qeUqqZLegjAyMhJjx47FkSNH0L17d2XoXb9+HTt37sTq1auxcOFCfZUneRZmJvBw/b+Lldzr10HrZvVx514RLufegZ21OVyd7OBc1wYA0Mz9////y7uH63n3la/HXb52Bxev5gF4eDj01y8n4/f9Z7D0mxTUq2MFAFBUCNy6U6DrKRJprKiwEJcuXVIuX/n7b/x15gxsbGzg7OICACgoKEBy8m+YNr3qkanKPpXMzc0BAA1cG6Kek5MOK6cn0VsQTpgwAQ4ODoiPj8fnn38OheLhBS6GhoZo164d1q1bh8GDB+urPMlr6+2G5K+mKJfnRw4CACT8fABj53yDvgGtsDrmLeX6hP+8DQD4ZNV2zPtiu1rbeL1HG9S1t8Lwfh0xvN//ndO9eDUPzftWf58WkT6dOnUSo8NHKpcXzn94S9BrIa9j7qcPz4f/tj0REAKv9umnlxpJczIhqrvVuXaVl5fj1q1bAAAHBwcYGxs/13hmbSZqoyyiF9qdtOX6LoFI70zV2N37R9xQb2xsDGdnZ32XQUREEsRHrBERkaQxCImISNIYhEREJGkMQiIikjQGIRERSRqDkIiIJI1BSEREksYgJCIiSWMQEhGRpDEIiYhI0hiEREQkaQxCIiKSNAYhERFJGoOQiIgkTeMgXL9+PRITE5XL77//PmxtbdGlSxdcvHhRq8URERHpmsZB+Omnn8LMzAwAkJqaihUrVmD+/PlwcHDA1KlTtV4gERGRLmn8xbyXL19GkyZNAABbtmzBoEGDMHbsWPj5+SEwMFDb9REREemUxnuElpaWyMvLAwAkJyejZ8+eAABTU1MUFxdrtzoiIiId03iPsGfPnhg9ejTatGmDc+fOoU+fPgCAU6dOwd3dXdv1ERER6ZTGe4QrVqxA586dcfPmTWzatAl16tQBABw5cgTDhg3TeoFERES6JBNCCH0XoW1mbSbquwQivbuTtlzfJRDpnakaxz3VOjR6/PhxtTfaunVrtfsSERHpm1pB6OvrC5lMhpp2HivXyWQyKBQKrRZIRESkS2oFYU5Ojq7rICIi0gu1gtDNzU3XdRAREenFMz1rNCEhAX5+fnBxcVE+Vm3x4sXYunWrVosjIiLSNY2DcOXKlYiIiECfPn1w9+5d5TlBW1tbLF68WNv1ERER6ZTGQbhs2TKsXr0aH330EQwNDZXt7du3x4kTJ7RaHBERka5pHIQ5OTlo06ZNlXa5XI7CwkKtFEVERFRbNA7CRo0aIT09vUr7b7/9Bi8vL23UREREVGs0ftZoREQEJkyYgJKSEgghcOjQIXz77beIjY3FV199pYsaiYiIdEbjIBw9ejTMzMwwa9YsFBUVYfjw4XBxccGSJUswdOhQXdRIRESkM8/1rNGioiIUFBSgbt262qzpufFZo0R81igRoMVnjVbnxo0bOHv2LICHj1hzdHR81qGIiIj0RuOLZe7fv4+33noLLi4uCAgIQEBAAFxcXDBixAjk5+frokYiIiKd0TgIR48ejYMHDyIxMRF3797F3bt3sW3bNhw+fBjvvPOOLmokIiLSGY3PEVpYWCApKQldu3ZVad+7dy969+79j7iXkOcIiXiOkAhQ7xyhxnuEderUgY2NTZV2Gxsb2NnZaTocERGRXmkchLNmzUJERARyc3OVbbm5uZg+fTpmz56t1eKIiIh0Ta2rRtu0aQOZTKZczszMRMOGDdGwYUMAwKVLlyCXy3Hz5k2eJyQioheKWkE4YMAAHZdBRESkH891Q/0/FS+WIeLFMkSAji6WISIieplo/GQZhUKB+Ph4fP/997h06RLKyspU1t++fVtrxREREemaxnuE0dHRWLRoEYYMGYL8/HxERERg4MCBMDAwQFRUlA5KJCIi0h2Ng3DDhg1YvXo1pk2bBiMjIwwbNgxfffUVPv74Yxw4cEAXNRIREemMxkGYm5uLVq1aAQAsLS2Vzxft168fEhMTtVsdERGRjmkchA0aNMC1a9cAAB4eHkhOTgYApKWlQS6Xa7c6IiIiHdM4CF9//XXs3LkTADBp0iTMnj0bTZs2xciRI/H2229rvUAiIiJdeu77CA8cOID9+/ejadOm6N+/v7bqei68j5CI9xESAbV0H+G//vUvREREoFOnTvj000+fdzgiIqJapbUny2RkZKBt27ZQKBTaGO65ZN8o1ncJRHoXNHeHvksg0rtLy157ah8+WYaIiCSNQUhERJLGICQiIklT+1mjERERT1x/8+bN5y6GiIiotqkdhMeOHXtqH39//+cqhoiIqLapHYS7du3SZR1ERER6wXOEREQkaQxCIiKSNAYhERFJGoOQiIgkjUFIRESS9kxBuHfvXowYMQKdO3fGlStXAAAJCQnYt2+fVosjIiLSNY2DcNOmTQgODoaZmRmOHTuG0tJSAEB+fj6/fYKIiF44GgfhJ598glWrVmH16tUwNjZWtvv5+eHo0aNaLY6IiEjXNA7Cs2fPVvsEGRsbG9y9e1cbNREREdUajYPQyckJWVlZVdr37duHxo0ba6UoIiKi2qJxEI4ZMwZTpkzBwYMHIZPJcPXqVWzYsAGRkZEYP368LmokIiLSGbWfNVrpgw8+QEVFBbp3746ioiL4+/tDLpcjMjISkyZN0kWNREREOiMTQohneWNZWRmysrJQUFAAb29vWFpaaru2Z5Z9o1jfJRDpXdDcHfougUjvLi177al9NN4jrGRiYgJvb+9nfTsREdE/gsZBGBQUBJlMVuP6lJSU5yqIiIioNmkchL6+virL5eXlSE9Px8mTJxEaGqqtuoiIiGqFxkEYHx9fbXtUVBQKCgqeuyAiIqLapLWHbo8YMQJr1qzR1nBERES1QmtBmJqaClNTU20NR0REVCs0PjQ6cOBAlWUhBK5du4bDhw9j9uzZWiuMiIioNmgchDY2NirLBgYG8PT0RExMDHr16qW1woiIiGqDRkGoUCgQHh6OVq1awc7OTlc1ERER1RqNzhEaGhqiV69e/JYJIiJ6aWh8sUzLli1x/vx5XdRCRERU657pi3kjIyOxbds2XLt2Dffu3VN5ERERvUjUPkcYExODadOmoU+fPgCA1157TeVRa0IIyGQyKBQK7VdJRESkI2oHYXR0NMaNG4ddu3bpsh4iIqJapXYQVn5bU0BAgM6KISIiqm0anSN80rdOEBERvYg0uo+wWbNmTw3D27dvP1dBREREtUmjIIyOjq7yZBkiIqIXmUZBOHToUNStW1dXtRAREdU6tc8R8vwgERG9jNQOwsqrRomIiF4mah8araio0GUdREREeqG1L+YlIiJ6ETEIiYhI0hiEREQkaQxCIiKSNAYhERFJGoOQiIgkjUFIRESSxiAkIiJJYxASEZGkMQiJiEjSGIRERCRpGn0NE0lX4ubvkbjlB1zPvQoAcGvkgWFhY9HhX10BAGWlpVi9Ig57diahvLwMbTt2wYSID2FnXwcAsGP7VsTHzql27I0/p8DWzr52JkL0HCzkhojs2xzBPs5wsJTj5N/5iNp0Escv3QUA9PZxxgg/N7RqaAs7CxP0/mw3Tl+5pzKGo5UcHw3wRtfmjrCUGyH7RgGWJ2Xi14xrepgRAQxCUpND3XoIHzcZLg0aQghg528/Y+7M97Bszf/g1qgJvly2EGmpezEzZgEsLC2xMv4zfPJRBOJWrgcA+HcPRrtOfipjxn/6McrKShmC9MKYP9wXns5WeO+/R3E9vxQDOzTAxomd0X3eLlzPL4G5iSHSzt/GtmNXMX+4b7VjxL/VBtbmxhj15SHcKShDSPv6+Pzt9ui34A+c+vtete8h3eKhUVJLJ78AdOjcDfVd3dCgoRtCx06CqZk5/jp1AoUF95GcuBljJk6Db7uOaOrpjakzo3HmZAb+OnUcACCXm8K+joPyZWhggIyjh9Cr7+t6nhmReuTGBnjVxxmfbj2NQ9m3cfFWIeJ/PYuLNwvxVld3AMBPaX9jyW/nsO/szRrHadfYHuv+yEHGxbu4lFeEZUmZuFdcjlautrUzEaqCQUgaUygU+OP331BSUgyvFq2RefYMHjx4AN/2nZR9XN0awbGeM86czKh2jJ1J2yA3NUXXoB61VTbRczEykMHI0ACl5apfSVdSrkAHD/WPahw5fxv927rAxtwYMhnQv60L5EYGSM3M03bJpKZ/9KHRy5cvY86cOVizZk2NfUpLS1FaWvpYWwXkcrmuy5OcnOxMTBs/EmVlZTAzM8PseYvQsJEHsrPOwsjYGJZW1ir97eztced29T/cSdu2ILDHq5DLTWujdKLnVliqwOHztzG5dzNk5d7HzfulCGnXAG0b2ePCzUK1x3l37WGsCG+PE/95FeWKChSXKTDmqzRcvKX+GKRd/+g9wtu3b2P9+vVP7BMbGwsbGxuV16qlC2qpQmlp0NAdy9d8h/gvEtAnZDDi5n2MSznZGo9z5mQGLl88j179eFiUXixTE45CBiBtXjCy4vshPLARth65ggoh1B5jWt/msDYzxrBl+9FvwR58tSsbn4e3h6ezle4KpyfS6x7hzz///MT158+ff+oYM2fOREREhErb3/kVNfSm52FsbAyXBg0BAE09vZH51yls/XEjur0SjAfl5Si4f09lr/DO7dvKq0YflbRtMxo39URTT+9aq51IGy7eKsLgpfthZmIIK1Mj3LhXihXh7XApr0it97s5mCM8oDF6zNuFc7n3AQBnrtxDR486CPVvhA+/O67L8qkGeg3CAQMGQCaTQTzhrymZTPbEMeRyeZXDoPKSYq3UR09WISpQXlaGpp5eMDIyQvqRQ+ga+PCc39+XLuDm9Wvwaumj8p7ioiLsTUlG2DuT9VEykVYUlylQXKaAjZkx/JvXRezW02q9z9TYEACq7EEqKgQMnvyrjnRIr4dGnZ2d8dNPP6GioqLa19GjR/VZHj1i7aqlOJF+BNevXUFOdubD5WOHEdirDywsrdCr7+tYvTwOGUfTkHn2NOJjP4ZXy9Zo3qK1yjh7UpKgUCgQ1KuPnmZC9Oz8mzsiwMsRrnXM0c3TEf+b3AXZ1+/j+wOXAAA25sbwrm+Npk4PD3N61LOEd31rOFo9/GM9+3oBcm4UIHaoD3zcbOHmYI4xr3igm6cjko7n6m1eUqfXPcJ27drhyJEjCAkJqXb90/YWqfbk372NuHmzcDvvFiwsLNHIoxnmxn2Oth06AwDGToqEzECGebOmoby8DO06dsG7ER9WGSc5cTO6BLxS5cIaoheBtZkxZvT3gpOtKfKLyrE94xoW/HIGDyoe/p7q2coJi0a0UfZfEd4eABC//Szifz2LBxUCoasO4oPXvLBmbCdYyA1x4VYhIr45hl2nb+hlTgTIhB6TZu/evSgsLETv3r2rXV9YWIjDhw8jICBAo3Gzb/DQKFHQ3B36LoFI7y4te+2pffS6R9itW7cnrrewsNA4BImIiDTxj759goiISNcYhEREJGkMQiIikjQGIRERSRqDkIiIJI1BSEREksYgJCIiSWMQEhGRpDEIiYhI0hiEREQkaQxCIiKSNAYhERFJGoOQiIgkjUFIRESSxiAkIiJJYxASEZGkMQiJiEjSGIRERCRpDEIiIpI0BiEREUkag5CIiCSNQUhERJLGICQiIkljEBIRkaQxCImISNIYhEREJGkMQiIikjQGIRERSRqDkIiIJI1BSEREksYgJCIiSWMQEhGRpDEIiYhI0hiEREQkaQxCIiKSNAYhERFJGoOQiIgkjUFIRESSxiAkIiJJYxASEZGkMQiJiEjSGIRERCRpDEIiIpI0BiEREUkag5CIiCSNQUhERJLGICQiIkljEBIRkaQxCImISNIYhEREJGkMQiIikjSZEELouwh6uZSWliI2NhYzZ86EXC7XdzlEesGfgxcHg5C07t69e7CxsUF+fj6sra31XQ6RXvDn4MXBQ6NERCRpDEIiIpI0BiEREUkag5C0Ti6XY86cObxAgCSNPwcvDl4sQ0REksY9QiIikjQGIRERSRqDkIiIJI1BSEREksYgJK1bsWIF3N3dYWpqik6dOuHQoUP6Lomo1uzZswf9+/eHi4sLZDIZtmzZou+S6CkYhKRV3333HSIiIjBnzhwcPXoUPj4+CA4Oxo0bN/RdGlGtKCwshI+PD1asWKHvUkhNvH2CtKpTp07o0KEDli9fDgCoqKiAq6srJk2ahA8++EDP1RHVLplMhs2bN2PAgAH6LoWegHuEpDVlZWU4cuQIevTooWwzMDBAjx49kJqaqsfKiIhqxiAkrbl16xYUCgXq1aun0l6vXj3k5ubqqSoioidjEBIRkaQxCElrHBwcYGhoiOvXr6u0X79+HU5OTnqqiojoyRiEpDUmJiZo164ddu7cqWyrqKjAzp070blzZz1WRkRUMyN9F0Avl4iICISGhqJ9+/bo2LEjFi9ejMLCQoSHh+u7NKJaUVBQgKysLOVyTk4O0tPTYW9vj4YNG+qxMqoJb58grVu+fDkWLFiA3Nxc+Pr6YunSpejUqZO+yyKqFbt370ZQUFCV9tDQUKxbt672C6KnYhASEZGk8RwhERFJGoOQiIgkjUFIRESSxiAkIiJJYxASEZGkMQiJiEjSGIRERCRpDEIiIpI0BiFRLQgLC1P5ctbAwEC89957tV7H7t27IZPJcPfuXZ1t4/G5PovaqJOoEoOQJCssLAwymQwymQwmJiZo0qQJYmJi8ODBA51v+6effsLcuXPV6lvboeDu7o7FixfXyraI/gn40G2StN69e2Pt2rUoLS3F9u3bMWHCBBgbG2PmzJlV+paVlcHExEQr27W3t9fKOET0/LhHSJIml8vh5OQENzc3jB8/Hj169MDPP/8M4P8O8c2bNw8uLi7w9PQEAFy+fBmDBw+Gra0t7O3tERISggsXLijHVCgUiIiIgK2tLerUqYP3338fjz/S9/FDo6WlpZgxYwZcXV0hl8vRpEkTfP3117hw4YLyAc52dnaQyWQICwsD8PArrmJjY9GoUSOYmZnBx8cHP/74o8p2tm/fjmbNmsHMzAxBQUEqdT4LhUKBUaNGKbfp6emJJUuWVNs3Ojoajo6OsLa2xrhx41BWVqZcp07tj7p48SL69+8POzs7WFhYoEWLFti+fftzzYWoEvcIiR5hZmaGvLw85fLOnTthbW2NHTt2AADKy8sRHByMzp07Y+/evTAyMsInn3yC3r174/jx4zAxMUFcXBzWrVuHNWvWwMvLC3Fxcdi8eTNeeeWVGrc7cuRIpKamYunSpfDx8UFOTg5u3boFV1dXbNq0CYMGDcLZs2dhbW0NMzMzAEBsbCy++eYbrFq1Ck2bNsWePXswYsQIODo6IiAgAJcvX8bAgQMxYcIEjB07FocPH8a0adOe6/OpqKhAgwYN8MMPP6BOnTrYv38/xo4dC2dnZwwePFjlczM1NcXu3btx4cIFhIeHo06dOpg3b55atT9uwoQJKCsrw549e2BhYYHTp0/D0tLyueZCpCSIJCo0NFSEhIQIIYSoqKgQO3bsEHK5XERGRirX16tXT5SWlirfk5CQIDw9PUVFRYWyrbS0VJiZmYmkpCQhhBDOzs5i/vz5yvXl5eWiQYMGym0JIURAQICYMmWKEEKIs2fPCgBix44d1da5a9cuAUDcuXNH2VZSUiLMzc3F/v37VfqOGjVKDBs2TAghxMyZM4W3t7fK+hkzZlQZ63Fubm4iPj6+xvWPmzBhghg0aJByOTQ0VNjb24vCwkJl28qVK4WlpaVQKBRq1f74nFu1aiWioqLUrolIE9wjJEnbtm0bLC0tUV5ejoqKCgwfPhxRUVHK9a1atVI5L5iRkYGsrCxYWVmpjFNSUoLs7Gzk5+fj2rVrKt+/aGRkhPbt21c5PFopPT0dhoaG1e4J1SQrKwtFRUXo2bOnSntZWRnatGkDADhz5kyV74Hs3Lmz2tuoyYoVK7BmzRpcunQJxcXFKCsrg6+vr0ofHx8fmJubq2y3oKAAly9fRkFBwVNrf9zkyZMxfvx4JCcno0ePHhg0aBBat2793HMhAnholCQuKCgIK1euhImJCVxcXGBkpPojYWFhobJcUFCAdu3aYcOGDVXGcnR0fKYaKg91aqKgoAAAkJiYiPr166usk8vlz1SHOv73v/8hMjIScXFx6Ny5M6ysrLBgwQIcPHhQ7TGepfbRo0cjODgYiYmJSE5ORmxsLOLi4jBp0qRnnwzR/8cgJEmzsLBAkyZN1O7ftm1bfPfdd6hbty6sra2r7ePs7IyDBw/C398fAPDgwQMcOXIEbdu2rbZ/q1atUFFRgT/++AM9evSosr5yj1ShUCjbvL29IZfLcenSpRr3JL28vJQX/lQ6cODA0yf5BH/++Se6dOmCd999V9mWnZ1dpV9GRgaKi4uVIX/gwAFYWlrC1dUV9vb2T629Oq6urhg3bhzGjRuHmTNnYvXq1QxC0gpeNUqkgX//+99wcHBASEgI9u7di5ycHOzevRuTJ0/G33//DQCYMmUKPvvsM2zZsgV//fUX3n333SfeA+ju7o7Q0FC8/fbb2LJli3LM77//HgDg5uYGmUyGbdu24ebNmygoKICVlRUiIyMxdepUrF+/HtnZ2Th69CiWLVuG9evXAwDGjRuHzMxMTJ8+HWfPnsXGjRuxbt06teZ55coVpKenq7zu3LmDpk2b4vDhw0hKSsK5c+cwe/ZspKWlVXl/WVkZRo0ahdOnT2P79u2YM2cOJk6cCAMDA7Vqf9x7772HpKQk5OTk4OjRo9i1axe8vLzUmgvRU+n7JCWRvjx6sYwm669duyZGjhwpHBwchFwuF40bNxZjxowR+fn5QoiHF8dMmTJFWFtbC1tbWxERESFGjhxZ48UyQghRXFwspk6dKpydnYWJiYlo0qSJWLNmjXJ9TEyMcHJyEjKZTISGhgohHl7gs3jxYuHp6SmMjY2Fo6OjCA4OFn/88Yfyfb/88oto0qSJkMvlolu3bmLNmjVqXSwDoMorISFBlJSUiLCwMGFjYyNsbW3F+PHjxQcffCB8fHyqfG4ff/yxqFOnjrC0tBRjxowRJSUlyj5Pq/3xi2UmTpwoPDw8hFwuF46OjuKtt94St27dqnEORJqQCVHDGXwiIiIJ4KFRIiKSNAYhERFJGoOQiIgkjUFIRESSxiAkIiJJYxASEZGkMQiJiEjSGIRERCRpDEIiIpI0BiEREUkag5CIiCTt/wH+z6ANavIL3AAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 500x300 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Assign\n",
    "cm = confusion_matrix(y_test, y_pred)\n",
    "\n",
    "\n",
    "# Plotting the heatmap\n",
    "plt.figure(figsize=(5, 3))  # Adjust size as needed\n",
    "sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', cbar=False)  # 'fmt' is a string format code\n",
    "plt.title('Confusion Matrix')\n",
    "plt.xlabel('Predicted Labels')\n",
    "plt.ylabel('True Labels')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ef0f9908-c22d-45a5-b4b5-95bb0a40b3fe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "miR: 0.0135\n",
      "expression: 0.0095\n",
      "reticulum: 0.0064\n",
      "assay: 0.0064\n",
      "cells: 0.0060\n",
      "YAP: 0.0058\n",
      "endoplasmic: 0.0055\n",
      "apoptosis: 0.0053\n",
      "Hippo: 0.0052\n",
      "ER: 0.0051\n",
      "invasion: 0.0050\n",
      "proliferation: 0.0049\n",
      "cell: 0.0047\n",
      "luciferase: 0.0046\n",
      "stress: 0.0044\n",
      "lines: 0.0040\n",
      "Western: 0.0039\n",
      "levels: 0.0039\n",
      "blot: 0.0039\n",
      "microRNA: 0.0035\n"
     ]
    }
   ],
   "source": [
    "# Get feature importances from the trained model\n",
    "importances = clf.feature_importances_\n",
    "\n",
    "# Get feature names from the vectorizer\n",
    "feature_names = tfidf_vectorizer.get_feature_names_out()\n",
    "\n",
    "# Define the number of top features you want to display\n",
    "K = 20\n",
    "\n",
    "# Get the indices of the K most important features in descending order of importance\n",
    "indices = importances.argsort()[-K:][::-1]\n",
    "\n",
    "# Display the top K important features with their names and importance scores\n",
    "top_features = [(feature_names[i], importances[i]) for i in indices]\n",
    "for feature, importance in top_features:\n",
    "    print(f\"{feature}: {importance:.4f}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
